### ISSUE TITLE
How to get the exception from a failed try?

### ISSUE THREAD

**Issue by raisercostin** [jasongoodwin/better-java-monads#22]
I couldn't find a way to get the exception from a failed try.

I was expecting to see something like
`Optional<RuntimeException> exception = Try.offFailable( ()-> throw new RuntimeException("a") ).getException()`

**Comment by Joe0** [comment]
That doesn't make any sense, as you don't know if a Try is failed or not. You can use the recover method or onFailure method and have access to the exception that way.

Iff you wanted to write bad code, as you're suggesting, you can always call the get method, catch the exception and have access to it that way. Though, doing what you want, or catching the exception defeats the entire purpose of this monad.

**Comment by raisercostin** [comment]
I will `Try` to bring some arguments to support my opinion:

- Following the first argument you wouldn't need a get/getUnchecked too since you don't know if a Try is failed or not. Nevertheless you can map a success, while you cannot map a failure in the current implementation.

- Scala offers the failed function to get the exception and also the pattern matching: http://www.scala-lang.org/api/2.9.3/scala/util/Failure.html .

- Try object transforms control structures in data structures which is great. If you have an exception (which is now encapsulated as data in a Failure) is not nice to be forced again to use control structures to get that data out.

- Consider the following case: you process something in a batch and for each item you could get an exception or a result: `List<Try<String>>` . Now from this structure you can't "recover" each exception but you would like to get a list of results `List<String>` and a list of the exceptions `List<Throwable>` to process/report them further.

Thanks for this library that is very useful.

**Comment by Joe0** [comment]
For this post, I'm going to refer to the try monad (what this library is implementing) as try. If I want to reference try, in general, or the keyword in Java, I will specify that.

> Following the first argument you wouldn't need a get/getUnchecked too since you don't know if a Try is failed or not. Nevertheless you can map a success, while you cannot map a failure in the current implementation.

You must not understand what get does... Get returns the value OR throws the exception. When you call it, you still don't know if it has failed or not. Also, the entire point of the try monad is for recoverable computations, meaning when everything is said and done, get should never throw an exception. Therefore, getException should not even exist, and if you do need it, you're misusing try.

> Scala offers the failed function to get the exception and also the pattern matching: http://www.scala-lang.org/api/2.9.3/scala/util/Failure.html . See http://stackoverflow.com/a/25588466/99248 for a usage and explanation of why this might be useful.

Java doesn't have pattern matching or extractor syntax. This is an argument that doesn't transfer to Java land. Furthermore, you shouldn't really be pattern matching on success/failure, that's the point of having recover and map. I'd also like to point out, Scala doesn't have a getException method either.

> Try object transforms control structures in data structures which is great. If you have an exception (which is now encapsulated as data in a Failure) is not nice to be forced again to use control structures to get that data out. 
A solution according to what you suggested would be

Please read what I wrote... I said do that if and only if you want to write bad code. I do not in any way endorse doing such things.

> Consider the following case: you process something in a batch and for each item you could get an exception or a result: List<Try<String>> . Now from this structure you can't "recover" each exception but you would like to get a list of results List<String> and a list of the exceptions List<Throwable> to process/report them further.

You're not using try correctly in that case. A batch operation is itself an atomic unit, and should itself be the try, not parts of it. The recover strategy for the batch operation might be retrying each sub-operation, if it would work. If you can't recover from a batch operation, then you shouldn't be using try.

**Comment by raisercostin** [comment]
Thanks for the replies. I will close the issue since the library works as designed.

**Comment by jluzonpr** [comment]
I tried to use this library today and had to avoid using it because there was no easy way of getting the exception without having to `catch` it. Scala has at the very least pattern matching to get it the `failed` method can be used.

In my case, I wanted to delegate the exception handling through more layers than the one it was generated in and wanted to avoid building an explicit class that wrapped it. I immediately thought I could use a `Try` and the library recommended this library, but IMO it is just incomplete. A simple accessor method for the exception that is available only on the `Failure` class, and the ability to cast my `Try` to either `Success` or `Failure` myself would be safe and still correct, so I'm not sure why at least this isn't a thing.

In the end, I used my own thing but thought I would add my feedback in here rather than just stay silent.

Cheers!
